File: chimera_api/routers/plan.py
import logging
from datetime import datetime, timedelta, date as date_type
from typing import Optional, Any
from uuid import UUID

from fastapi import APIRouter, Body, Depends, HTTPException, Query
from dependencies import get_current_user
from schemas import (
    PlanImportRequest,
    PhaseCreate,
    PhaseUpdate,
    TemplateCreate,
    MoveWorkoutRequest,
    DuplicateWorkoutRequest,
    WeekActionRequest,
    SaveWeekTemplateRequest,
    ApplyTemplateRequest,
)
from db_client import supabase_admin
from services import plan_action_service, phase_service, template_service, plan_import_service

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/v1/plan", tags=["Plan"])


# --- Plan Import (flexible format) ---

@router.post("/import")
async def import_plan(
    raw_input: Any = Body(...),
    user_id: str = Depends(get_current_user),
):
    """
    Flexible plan import accepting multiple shapes:
    - Shape A: bare array of entries, e.g. [{date, title, ...}, ...]
    - Shape B: {entries: [...], distance_unit?: str}
    - Shape C: system format {workouts: [...], phases?: [...]}
    """
    try:
        return await plan_import_service.import_plan(user_id, raw_input)
    except Exception as e:
        logger.error(f"Plan import failed: {e}")
        raise HTTPException(status_code=500, detail=f"Import failed: {e}")


# --- Plan Export ---

@router.get("/export")
async def export_plan(
    start_date: str = Query(..., description="YYYY-MM-DD"),
    end_date: str = Query(..., description="YYYY-MM-DD"),
    user_id: str = Depends(get_current_user),
):
    """Export plan as system format JSON for a date range."""
    try:
        sd = date_type.fromisoformat(start_date)
        ed = date_type.fromisoformat(end_date)
    except ValueError:
        raise HTTPException(status_code=400, detail="Invalid date format. Use YYYY-MM-DD.")
    return await plan_import_service.export_plan(user_id, sd, ed)


# --- Combined Calendar Data ---

@router.get("/calendar")
async def get_calendar_data(
    start_date: str = Query(..., description="YYYY-MM-DD"),
    weeks: int = Query(5, ge=1, le=12),
    user_id: str = Depends(get_current_user),
):
    """Single fetch for the web training plan view."""
    return await plan_action_service.get_calendar_data(user_id, start_date, weeks)


# --- Workout Actions ---

@router.put("/workouts/{workout_id}/move")
async def move_workout(
    workout_id: UUID,
    body: MoveWorkoutRequest,
    user_id: str = Depends(get_current_user),
):
    return await plan_action_service.move_workout(str(workout_id), body.new_date, user_id)


@router.post("/workouts/{workout_id}/duplicate")
async def duplicate_workout(
    workout_id: UUID,
    body: DuplicateWorkoutRequest,
    user_id: str = Depends(get_current_user),
):
    return await plan_action_service.duplicate_workout(str(workout_id), body.target_date, user_id)


@router.delete("/workouts/{workout_id}")
async def delete_workout(
    workout_id: UUID,
    user_id: str = Depends(get_current_user),
):
    return await plan_action_service.delete_workout(str(workout_id), user_id)


# --- Week Actions ---

@router.put("/weeks/move")
async def move_week(
    body: WeekActionRequest,
    user_id: str = Depends(get_current_user),
):
    return await plan_action_service.move_week(
        body.source_week_start, body.target_week_start, user_id
    )


@router.post("/weeks/duplicate")
async def duplicate_week(
    body: WeekActionRequest,
    user_id: str = Depends(get_current_user),
):
    return await plan_action_service.duplicate_week(
        body.source_week_start, body.target_week_start, user_id
    )


@router.delete("/weeks/{week_start}/clear")
async def clear_week(
    week_start: date_type,
    user_id: str = Depends(get_current_user),
):
    return await plan_action_service.clear_week(week_start, user_id)


@router.post("/weeks/save-template")
async def save_week_as_template(
    body: SaveWeekTemplateRequest,
    user_id: str = Depends(get_current_user),
):
    return await plan_action_service.save_week_as_template(
        body.week_start, body.title, user_id
    )


# --- Phase CRUD ---

@router.get("/phases")
async def get_phases(
    start_date: Optional[str] = None,
    end_date: Optional[str] = None,
    user_id: str = Depends(get_current_user),
):
    return await phase_service.get_phases(user_id, start_date, end_date)


@router.post("/phases")
async def create_phase(
    body: PhaseCreate,
    user_id: str = Depends(get_current_user),
):
    return await plan_action_service.create_phase(
        body.model_dump(exclude_unset=True), user_id
    )


@router.patch("/phases/{phase_id}")
async def update_phase(
    phase_id: UUID,
    body: PhaseUpdate,
    user_id: str = Depends(get_current_user),
):
    return await plan_action_service.update_phase(
        str(phase_id), body.model_dump(exclude_unset=True), user_id
    )


@router.delete("/phases/{phase_id}")
async def delete_phase(
    phase_id: UUID,
    user_id: str = Depends(get_current_user),
):
    return await plan_action_service.delete_phase(str(phase_id), user_id)


# --- Template CRUD + Apply ---

@router.get("/templates")
async def get_templates(
    type: Optional[str] = Query(None, alias="type"),
    user_id: str = Depends(get_current_user),
):
    return await template_service.get_templates(user_id, type)


@router.post("/templates")
async def create_template(
    body: TemplateCreate,
    user_id: str = Depends(get_current_user),
):
    return await template_service.create_template(
        body.model_dump(), user_id
    )


@router.delete("/templates/{template_id}")
async def delete_template(
    template_id: UUID,
    user_id: str = Depends(get_current_user),
):
    await template_service.delete_template(str(template_id), user_id)
    return {"status": "deleted"}


@router.post("/templates/{template_id}/apply")
async def apply_template(
    template_id: UUID,
    body: ApplyTemplateRequest,
    user_id: str = Depends(get_current_user),
):
    return await plan_action_service.apply_template(
        str(template_id), body.start_date, body.detail_level, user_id
    )


# --- Agent Action Log ---

@router.get("/agent-actions")
async def get_agent_actions(
    limit: int = Query(20, ge=1, le=100),
    user_id: str = Depends(get_current_user),
):
    return await plan_action_service.get_agent_actions(user_id, limit)


@router.post("/agent-actions/{action_id}/revert")
async def revert_agent_action(
    action_id: UUID,
    user_id: str = Depends(get_current_user),
):
    return await plan_action_service.revert_agent_action(str(action_id), user_id)

File: chimera_mobile_app/src/domain/api/plan.ts
import type { FetchFn } from './client';
import type { CalendarData, TrainingPhase, PlanTemplate, AgentAction } from '../types/plan';

export async function getCalendarData(
  fetch: FetchFn,
  startDate: string,
  weeks: number = 5
): Promise<CalendarData> {
  const res = await fetch(`/plan/calendar?start_date=${startDate}&weeks=${weeks}`);
  if (!res.ok) throw new Error(`Failed to fetch calendar data: ${res.status}`);
  return res.json();
}

export async function moveWorkout(
  fetch: FetchFn,
  workoutId: string,
  newDate: string
): Promise<any> {
  const res = await fetch(`/plan/workouts/${workoutId}/move`, {
    method: 'PUT',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ new_date: newDate }),
  });
  if (!res.ok) throw new Error(`Failed to move workout: ${res.status}`);
  return res.json();
}

export async function duplicateWorkout(
  fetch: FetchFn,
  workoutId: string,
  targetDate: string
): Promise<any> {
  const res = await fetch(`/plan/workouts/${workoutId}/duplicate`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ target_date: targetDate }),
  });
  if (!res.ok) throw new Error(`Failed to duplicate workout: ${res.status}`);
  return res.json();
}

export async function deleteWorkout(
  fetch: FetchFn,
  workoutId: string
): Promise<any> {
  const res = await fetch(`/plan/workouts/${workoutId}`, {
    method: 'DELETE',
  });
  if (!res.ok) throw new Error(`Failed to delete workout: ${res.status}`);
  return res.json();
}

export async function moveWeek(
  fetch: FetchFn,
  sourceWeekStart: string,
  targetWeekStart: string
): Promise<any> {
  const res = await fetch('/plan/weeks/move', {
    method: 'PUT',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ source_week_start: sourceWeekStart, target_week_start: targetWeekStart }),
  });
  if (!res.ok) throw new Error(`Failed to move week: ${res.status}`);
  return res.json();
}

export async function duplicateWeek(
  fetch: FetchFn,
  sourceWeekStart: string,
  targetWeekStart: string
): Promise<any> {
  const res = await fetch('/plan/weeks/duplicate', {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ source_week_start: sourceWeekStart, target_week_start: targetWeekStart }),
  });
  if (!res.ok) throw new Error(`Failed to duplicate week: ${res.status}`);
  return res.json();
}

export async function clearWeek(
  fetch: FetchFn,
  weekStart: string
): Promise<any> {
  const res = await fetch(`/plan/weeks/${weekStart}/clear`, {
    method: 'DELETE',
  });
  if (!res.ok) throw new Error(`Failed to clear week: ${res.status}`);
  return res.json();
}

export async function saveWeekAsTemplate(
  fetch: FetchFn,
  weekStart: string,
  title: string
): Promise<PlanTemplate> {
  const res = await fetch('/plan/weeks/save-template', {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ week_start: weekStart, title }),
  });
  if (!res.ok) throw new Error(`Failed to save template: ${res.status}`);
  return res.json();
}

export async function getPhases(fetch: FetchFn): Promise<TrainingPhase[]> {
  const res = await fetch('/plan/phases');
  if (!res.ok) throw new Error(`Failed to fetch phases: ${res.status}`);
  return res.json();
}

export async function createPhase(
  fetch: FetchFn,
  data: Partial<TrainingPhase>
): Promise<TrainingPhase> {
  const res = await fetch('/plan/phases', {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(data),
  });
  if (!res.ok) throw new Error(`Failed to create phase: ${res.status}`);
  return res.json();
}

export async function updatePhase(
  fetch: FetchFn,
  phaseId: string,
  data: Partial<TrainingPhase>
): Promise<TrainingPhase> {
  const res = await fetch(`/plan/phases/${phaseId}`, {
    method: 'PATCH',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(data),
  });
  if (!res.ok) throw new Error(`Failed to update phase: ${res.status}`);
  return res.json();
}

export async function deletePhase(
  fetch: FetchFn,
  phaseId: string
): Promise<void> {
  const res = await fetch(`/plan/phases/${phaseId}`, { method: 'DELETE' });
  if (!res.ok) throw new Error(`Failed to delete phase: ${res.status}`);
}

export async function getTemplates(
  fetch: FetchFn,
  type?: string
): Promise<PlanTemplate[]> {
  const url = type ? `/plan/templates?type=${type}` : '/plan/templates';
  const res = await fetch(url);
  if (!res.ok) throw new Error(`Failed to fetch templates: ${res.status}`);
  return res.json();
}

export async function createTemplate(
  fetch: FetchFn,
  data: Partial<PlanTemplate>
): Promise<PlanTemplate> {
  const res = await fetch('/plan/templates', {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(data),
  });
  if (!res.ok) throw new Error(`Failed to create template: ${res.status}`);
  return res.json();
}

export async function deleteTemplate(
  fetch: FetchFn,
  templateId: string
): Promise<void> {
  const res = await fetch(`/plan/templates/${templateId}`, { method: 'DELETE' });
  if (!res.ok) throw new Error(`Failed to delete template: ${res.status}`);
}

export async function applyTemplate(
  fetch: FetchFn,
  templateId: string,
  startDate: string,
  detailLevel: 'full' | 'structure' = 'full'
): Promise<any> {
  const res = await fetch(`/plan/templates/${templateId}/apply`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ start_date: startDate, detail_level: detailLevel }),
  });
  if (!res.ok) throw new Error(`Failed to apply template: ${res.status}`);
  return res.json();
}

export async function getAgentActions(
  fetch: FetchFn,
  limit: number = 20
): Promise<AgentAction[]> {
  const res = await fetch(`/plan/agent-actions?limit=${limit}`);
  if (!res.ok) throw new Error(`Failed to fetch agent actions: ${res.status}`);
  return res.json();
}

export async function revertAgentAction(
  fetch: FetchFn,
  actionId: string
): Promise<any> {
  const res = await fetch(`/plan/agent-actions/${actionId}/revert`, {
    method: 'POST',
  });
  if (!res.ok) throw new Error(`Failed to revert action: ${res.status}`);
  return res.json();
}
